rust/kernel/usb.rs

```diff
@@ -0,0 +1,421 @@
+            bInterfaceNumber: self.b_interface_number,
+            driver_info: info as _,
+        }
+    }
+}
+
+/// Define a const USB device id table
+///
+/// # Examples
+///
+/// ```ignore
+/// # use kernel::{usb, define_usb_id_table};
+/// #
+/// struct MyDriver;
+/// impl usb::Driver for MyDriver {
+///     // [...]
+/// #   fn probe(_dev: &mut usb::Device, _id_info: Option<&Self::IdInfo>) -> Result {
+/// #       Ok(())
+/// #   }
+/// #   define_usb_id_table! {u64, [
+/// #       (usb::DeviceId::new(0xbee7, 0xffff), None),
+/// #       (usb::DeviceId::with_class(0xbee7, 0xff1d), Some(0x40)),
+/// #   ]}
+/// }
+#[macro_export]
+macro_rules! define_usb_id_table {
+    ($data_type:ty, $($t:tt)*) => {
+        type IdInfo = $data_type;
+        const ID_TABLE: $crate::driver::IdTable<'static, $crate::usb::DeviceId, $data_type> = {
+            $crate::define_id_array!(ARRAY, $crate::usb::DeviceId, $data_type, $($t)* );
+            ARRAY.as_table()
+        };
+    };
+}
+
+/// An USB driver
+pub trait Driver {
+    /// Data stored on device interface by driver.
+    ///
+    /// Corresponds to the data set or retrieved via the kernel's
+    /// `usb_{set,get}_intfdata()` functions.
+    ///
+    /// Require that `Data` implements `PointerWrapper`. We guarantee to
+    /// never move the underlying wrapped data structure.
+    type Data: PointerWrapper + Send + Sync + driver::DeviceRemoval = ();
+
+    /// The type holding information about each device id supported by the driver.
+    /// Trait `From<NonZeroU64>` is required due to driver info being either a
+    /// flagset or a pointer to driver-specific data.
+    type IdInfo: 'static + From<NonZeroU64> = u64;
+
+    /// The table of device ids supported by the driver.
+    const ID_TABLE: driver::IdTable<'static, DeviceId, Self::IdInfo>;
+
+    /// USB driver probe.
+    ///
+    /// Called to see if the driver can manage a device interface.
+    /// Implementer should attempt to initialize the interface here.
+    fn probe(dev: &mut Device, id: Option<&Self::IdInfo>) -> Result<Self::Data>;
+
+    /// USB driver disconnect.
+    ///
+    /// Called when a device interface is removed.
+    /// Implementer should prepare the interface for removal here.
+    fn disconnect(data: &Self::Data);
+}
+
+/// An USB device.
+///
+/// # Invariants
+///
+/// The field `ptr` is non-null and valid for the lifetime of the object.
+pub struct Device {
+    ptr: *mut bindings::usb_device,
+}
+
+impl Device {
+    unsafe fn from_ptr(ptr: *mut bindings::usb_device) -> Self {
+        Self { ptr }
+    }
+}
+
+unsafe impl device::RawDevice for Device {
+    fn raw_device(&self) -> *mut bindings::device {
+        // SAFETY: By the type invariants, we know that `self.ptr` is non-null and valid.
+        unsafe { &mut (*self.ptr).dev }
+    }
+}
+
+/// An USB device interface.
+///
+/// # Invariants
+///
+/// The field `ptr` is non-null and valid for the lifetime of the object.
+pub struct Interface {
```

Comment:
Could we include an example about how this is expected to be used?

Reply:
I can't decide whether to include an example on this PR or on a different one, specially since the hardware that I use have _quite sized_ drivers but I'll see what can I do.

Reply:
It could just be something trivial like an echo server. But ideally we also implement some additional safe wrappers around the other types that need to be used (`urb`s, etc.) so that it can be written without almost any `unsafe` code.

Reply:
Done, after delving in the source code of the kernel I've decided to create a simple example based on `usb-serial` but only processes bulk-in URBs and puts the transfer data on a misc device.